File: ArithmeticSlicesII-Subsequence2.py
class Solution:
    def numberOfArithmeticSlices(self, A):
        """
        :type A: List[int]
        :rtype: int
        """
        n = len(A)
        needDiffCount = {}
        s = A.copy()
        result = 0
        for i in range(1, n):
            if A[i] in needDiffCount:
                for key, value in needDiffCount[A[i]].items():
                    need = A[i] + key
                    result += value
                    if need in s:
                        if need not in needDiffCount:
                            needDiffCount[need] = {}
                        needDiffCount[need][key] = needDiffCount[need].get(key, 0) + value
            for j in range(i):
                diff = A[i] - A[j]
                need = A[i] + diff
                if need in s:
                    if need not in needDiffCount:
                        needDiffCount[need] = {}
                    needDiffCount[need][diff] = needDiffCount[need].get(diff, 0) + 1
            s.remove(A[i])
        return result


if __name__ == '__main__':
    A = [int(num) for num in input().split()]
    solution = Solution()
    print(solution.numberOfArithmeticSlices(A))

File: ArithmeticSlicesII-Subsequence2.cpp
#include<iostream>
#include<vector>
#include<unordered_map>
#include<algorithm>

using namespace std;

class Solution {
public:
    int numberOfArithmeticSlices(vector<int>& A) {
        vector<unordered_map<int, int> > dp(A.size());
        int result = 0;
        for(int i = 0; i < A.size(); i++) {
        	for(int j = 0; j < i; j++) {
        		long long delta = (long long)A[i] - A[j];
        		if(delta > INT_MAX || delta < INT_MIN) continue;
        		int diff = (int)delta;
        		dp[i][diff]++;
        		if(dp[j].find(diff) != dp[j].end()) {
        			dp[i][diff] += dp[j][diff];
        			result += dp[j][diff];
				}
			}
		}
		
		return result;
    }
};

int main() {
	
	int n;
	cin>>n;
	
	vector<int> nums;
	for(int i = 0; i < n; i++) {
		int num;
		cin>>num;
		nums.push_back(num);
	} 
	
	Solution *solution = new Solution();
	cout<<solution->numberOfArithmeticSlices(nums);
	
	return 0;
	
}
